File: backend/app/schemas/article.py
from pydantic import BaseModel
from datetime import datetime
from app.schemas.tag import TagOut
from app.schemas.user import UserOut


class ArticleAuthor(BaseModel):
    id: int
    username: str
    avatar_url: str | None

    model_config = {"from_attributes": True}


class ArticleBase(BaseModel):
    title: str
    slug: str | None = None
    summary: str | None = None
    content: str = ""
    cover_image: str | None = None
    status: str = "draft"
    is_featured: bool = False
    tag_ids: list[int] = []
    published_at: datetime | None = None


class ArticleCreate(ArticleBase):
    pass


class ArticleUpdate(BaseModel):
    title: str | None = None
    slug: str | None = None
    summary: str | None = None
    content: str | None = None
    cover_image: str | None = None
    status: str | None = None
    is_featured: bool | None = None
    tag_ids: list[int] | None = None
    published_at: datetime | None = None


class ArticleListItem(BaseModel):
    id: int
    title: str
    slug: str
    summary: str | None
    cover_image: str | None
    author: ArticleAuthor | None
    status: str
    is_featured: bool
    view_count: int
    read_time: int
    tags: list[TagOut]
    published_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ArticleOut(ArticleListItem):
    content: str
    updated_at: datetime

File: frontend/src/types/article.ts
import type { Tag } from './tag'
import type { User } from './auth'

export type ArticleStatus = 'draft' | 'published'

export interface Article {
  id: number
  title: string
  slug: string
  summary?: string
  content: string
  cover_image?: string
  author?: User
  status: ArticleStatus
  is_featured: boolean
  view_count: number
  read_time: number
  tags: Tag[]
  published_at?: string
  created_at: string
  updated_at: string
}

export interface ArticleListItem {
  id: number
  title: string
  slug: string
  summary?: string
  cover_image?: string
  author?: Pick<User, 'id' | 'username' | 'avatar_url'>
  status: ArticleStatus
  is_featured: boolean
  view_count: number
  read_time: number
  tags: Tag[]
  published_at?: string
  created_at: string
}

export interface ArticleCreate {
  title: string
  slug?: string
  summary?: string
  content: string
  cover_image?: string
  status?: ArticleStatus
  is_featured?: boolean
  tag_ids?: number[]
  published_at?: string
}

export interface ArticleUpdate extends Partial<ArticleCreate> {}

export interface PaginatedArticles {
  items: ArticleListItem[]
  total: number
  page: number
  size: number
  pages: number
}
